adr/ADR-15.md

```diff
@@ -0,0 +1,137 @@
+
+## Design
+
+### Creation
+
+The library should have API(s) that allow(s) creation of a JetStream subscription. Depending on the language, there may be various type of subscriptions (like Channel based in Go), or with various options such as Queue subscriptions, etc... There can be different APIs or one with a set of options.
+
+#### Configuration
+
+When creating a subscription, the user can provide several things:
+
+- A subject (see meaning of this below)
+- The stream name
+- The consumer name, really a durable name for when the JS consumer already exists
+- A Queue name if this subscription is meant to be a queue subscription
+- A JS consumer configuration (see below), if the library is expected to create the JS consumer
+- Indication if this is a "Pull" subscription or not
+
+At the time of this writing, the consumer configuration object looks like this:
+
+```go
+type ConsumerConfig struct {
+	Durable         string        `json:"durable_name,omitempty"`
+	DeliverSubject  string        `json:"deliver_subject,omitempty"`
+	DeliverPolicy   DeliverPolicy `json:"deliver_policy"`
+	OptStartSeq     uint64        `json:"opt_start_seq,omitempty"`
+	OptStartTime    *time.Time    `json:"opt_start_time,omitempty"`
+	AckPolicy       AckPolicy     `json:"ack_policy"`
+	AckWait         time.Duration `json:"ack_wait,omitempty"`
+	MaxDeliver      int           `json:"max_deliver,omitempty"`
+	FilterSubject   string        `json:"filter_subject,omitempty"`
+	ReplayPolicy    ReplayPolicy  `json:"replay_policy"`
+	RateLimit       uint64        `json:"rate_limit_bps,omitempty"` // Bits per sec
+	SampleFrequency string        `json:"sample_freq,omitempty"`
+	MaxWaiting      int           `json:"max_waiting,omitempty"`
+	MaxAckPending   int           `json:"max_ack_pending,omitempty"`
+	FlowControl     bool          `json:"flow_control,omitempty"`
+	Heartbeat       time.Duration `json:"idle_heartbeat,omitempty"`
+}
+```
+
+#### Stream name
+
+When no stream name is specified, the library will use the subject provided as a way to find out which stream this subscription is for. A request is sent to the server on the `<JS prefix>.STREAM.NAMES` subject with a JSON content `{"subject":"<subject here>"}`. If the response (a list of stream names) is positive and contains a single entry, then the library will use this stream name, otherwise an error indicating that there is no matching stream name should be returned.
+
+#### Consumer name
+
+When no consumer name is specified, but a configuration is provided, the consumer will be the `Durable` field of the `ConsumerConfig` object.
+
+Note if the user provides a `Durable` in the configuration, this is treated differently that if the user provides a consumer to the subscribe API. Providing the stream and consumer name through the API makes the subscription being "bound", that is, no attempt to create a JS consumer will be done because the user expects the stream and consumer to already exists.
```

Comment:
Change requested:
There is a separate API call `Bind` which requires a stream name and a durable consumer name. When binding, the subscription will be created with the assumption that both the stream and the consumer already exist.